cache/edge_lambdas/src/rejection.ts

```diff
@@ -0,0 +1,205 @@
+ * of accepted requests looked like bona-fide traffic.
+ *
+ */
+
+import {
+  CloudFrontRequest,
+  CloudFrontRequestEvent,
+  CloudFrontResponse,
+} from 'aws-lambda';
+
+// Returns true if this request looks like a bot.
+//
+// A lot of spam requests will spoof their User-Agent header to look like a popular
+// web crawler, e.g. Googlebot or Bingbot.  Because a person would never spoof their
+// User-Agent this way, we can use this as an excuse to apply more stringent checks.
+const isBotRequest = (request: CloudFrontRequest): boolean => {
+  const userAgent = request.headers['user-agent']?.[0]?.value || '';
+
+  return (
+    userAgent.toLowerCase().includes('bot') ||
+    userAgent.toLowerCase().includes('search.yahoo')
+  );
+};
+
+// 253,376 requests rejected
+// 6,688,393 requests accepted
+
+export const looksLikeSpam = (event: CloudFrontRequestEvent): boolean => {
+  const cf = event.Records[0].cf;
+  const request = cf.request;
+
+  const requestParams = new URLSearchParams(request.querystring);
+  const query = requestParams.getAll('query').join(' ');
+
+  // Look for the `query` URL parameter.
+  //
+  // The majority of our spam comes from people plugging in values to the
+  // search form and stuffing bogus values into the query parameter.
+  // If there isn't a query parameter on this request, it's probably not spam.
+  if (query === '') {
+    return false;
+  }
+
+  // Look for spammy keywords and emoji.
+  //
+  // We can't solely rely on the presence of these words/characters to indicate spam,
+  // but they're a clue -- usually accompanied by long strings of Chinese characters
+  // and a URL to a sketchy-looking site.
+  //
+  // To avoid penalising real users, we only treat this text as spam if the user
+  // has self-identified as a bot/crawler in their User-Agent header.
+  //
+  // In my analysis, this flagged ~41% of all search traffic.
+  const spamKeywords = [
+    'cash',
+    'casino',
+    'chatgpt',
+    'crypto',
+    'ddos',
+    'free',
+    'fun',
+    'game',
+    'give away',
+    'give-away',
+    'giveaway',
+    'play',
+    'poker',
+    'telegram',
+    'whatsapp',
+    'win',
+    '➡️',
+    '☀️',
+    '⏩',
+    '⛔',
+    '⚽',
+    '✅',
+    '🐯',
+    '🍀',
+    '㊙',
+    '✔️',
+    '⭐',
+    '👈',
+    '👉',
+  ];
+
+  if (
+    isBotRequest(request) &&
+    spamKeywords.some(kw => query.toLowerCase().includes(kw))
+  ) {
+    return true;
+  }
+
+  // Look for query parameters that we don't use.
+  //
+  // I noticed that a lot of search requests include query parameters that we don't
+  // define in our apps, and with values that aren't meaningful.  If one of these are
+  // present, it suggests you're probably not a real person.
+  //
+  // In my analysis, this flagged ~4% of all search traffic.
+  const unrecognisedParams = ['ID', 'news', 'action', 'cat', 'token', 'type'];
+
+  if (unrecognisedParams.some(name => requestParams.get(name) !== null)) {
+    return true;
+  }
```

Comment:
Is this one high enough value / how much of that 4% lives within the bins defined by the other heuristics? Might lean towards getting rid of this

Reply:
This was the first check I wrote, because compared to the "unusual character" check it's pretty fast – but that value would be stronger if I'd made it the first check that runs, rather that burying it behind the spam keyword check.

As for stats:

```
# with the check
5,319,191 requests rejected
1,622,578 requests accepted
Takes ~1.7021207984995127e-05s to run
0.7662587158979217

# without the check
5,191,844 requests rejected
1,749,925 requests accepted
Takes ~2.14876845334318e-05s to run
0.7479136802160947
```

I reckon we could lose it.

Reply:
Removed.